Channel and author check/Inventory subcommands

KatLands opened the pull request:
Added a channel and author check to resolve an earlier issue. 
Added inventory subcommands.

Bot will now check if the message is in the channel as the one it was originally called from and also checks that the author is the same as the one who made the original command call.

New inventory commands are:
!inv
!inv add
!inv remove

KatLands committed:
Channel and author check/Inventory sub-commands

Added a channel and author check to resolve an earlier issue. Added inventory sub-commands.

Bot will now check if the message is in the channel as the one it was originally called from and also checks that the author is the same as the one who made the original command call.

New inventory commands are:
!inv
!inv add
!inv remove

jwhett commented on bot.py:
```diff
@@ -112,7 +112,28 @@ async def list(ctx):
+                    "fields": [
+                        {
+                            "name": f"__*{name}'s Inventory:*__",
+                            "value": one_line,
```
No spaces around `=` during a function call. I think these names are more descriptive:

- `author_id`: `author` suggests to me that I probably have a `User` object where in fact it is only the ID.
- `author_name`
- `one_per_line`

I would be happy collapsing most of it since we only use each exactly once and we don't lose any clarity. However, empty inventories will fail completely as `value` becomes an empty string and that's not allowed (`HTTPException: 400 Bad Request`) so we have to guard against that at least when we try to report an empty inventory back to the author.

```suggestion
        one_per_line = "\n".join(tracker.get_inv(ctx.message.author.id))
        if len(one_per_line) == 0:
            one_per_line = "<< Empty >>"
        await ctx.message.channel.send(
            embed = Embed().from_dict(
                {
                    "fields": [
                        {
                            "name": f"__*{ctx.message.author.name}'s Inventory:*__",
                            "value": one_per_line,
```

KatLands replied:
Suggested changes made. Thanks!

jwhett commented on bot.py:
```diff
@@ -127,12 +148,19 @@ async def add(ctx):
-
-
-@bot.command()
+    while True:
+        try:
+            add_message = await bot.wait_for("message", timeout=60)   
+            if ctx.message.channel.id == add_message.channel.id and author == add_message.author.id:
+                tracker.add_inv(author, add_message.content)
+                await add_message.add_reaction("✅")
+                break
+        except:
+            await ctx.message.channel.send(f"Inventory add request timed out. Please use {bot_prefix}inv add again.")
+            break
```
Since [`bot.wait_for()`](https://discordpy.readthedocs.io/en/stable/ext/commands/api.html?highlight=command%20ext%20bot#discord.ext.commands.Bot.wait_for) just raises an [`asyncio.TimeoutError`](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.TimeoutError) when the timeout occurs, we can just catch that. The suggestion below is dependent on adding the following import: `from asyncio import TimeoutError`

```suggestion
    while True:
        try:
            add_message = await bot.wait_for("message", timeout=60)
        except TimeoutError:
            await ctx.message.channel.send(
                f"Inventory add request timed out. Please use `{bot_prefix}inv` add again."
            )
            break

        if (
            ctx.message.channel.id == add_message.channel.id
            and author == add_message.author.id
        ):
            tracker.add_inv(author, add_message.content)
            await add_message.add_reaction("✅")
            break
```

I have the same thoughts about `author` in this case as in my prior comment. I believe it should be `author_id` if it is the author's ID. Though, this is outside of the changes in this function and should be refactored later.

KatLands replied:
Suggested changes made. Thanks!

jwhett commented on bot.py:
```diff
@@ -147,28 +175,16 @@ async def remove(ctx):
-            }
-        )
-    )
+    while True:
+        try:
+            remove_message = await bot.wait_for("message", timeout=60)
+            if ctx.message.channel.id == remove_message.channel.id and author == remove_message.author.id:
+                tracker.remove_inv(author, remove_message.content)
+                await remove_message.add_reaction("❎")
+                break
+        except:
+            await ctx.message.channel.send(f"Inventory remove request timed out. Please use {bot_prefix}inv remove again.")
+            break
```
Same as `!inv add`, we can rely on `asyncio.TimeoutError` to simplify the code a bit.

```suggestion
    while True:
        try:
            remove_message = await bot.wait_for("message", timeout=60)
        except TimeoutError:
            await ctx.message.channel.send(
                f"Inventory remove request timed out. Please use `{bot_prefix}inv` add again."
            )
            break

        remove_message = await bot.wait_for("message", timeout=60)
        if (
            ctx.message.channel.id == remove_message.channel.id
            and author == remove_message.author.id
        ):
            tracker.remove_inv(author, remove_message.content)
            await remove_message.add_reaction("❎")
            break
```

jwhett replied:
Also same comments here with `author` vs `author_id`.

KatLands replied:
Suggested changes made. Thanks!

jwhett reviewed (changes requested):
There are ways to simplify the error handling around timeouts and some formatting to clean up, but the feature works!

Have to guard against the cases of empty inventories though. I was getting HTTP Errors when sending empty strings (empty inventories) as `value` in an Embed.

KatLands committed:
Update to Channel and author check/Inventory subcommands

This is an update to the previous commit with the suggested changes from jwhett implemented.

Added new import for TimeourError
Adjusted try catch
Added a channel and author check to resolve an earlier issue.
Added inventory subcommands.

Bot will now check if the message is in the channel as the one it was originally called from and also checks that the author is the same as the one who made the original command call.

New inventory commands are:
!inv
!inv add
!inv remove
